File: lev1/두 정수 사이의 합.py
"""
문제 설명
두 정수 a, b가 주어졌을 때 a와 b 사이에 속한 모든 정수의 합을 리턴하는 함수, solution을 완성하세요.
예를 들어 a = 3, b = 5인 경우, 3 + 4 + 5 = 12이므로 12를 리턴합니다.

제한 조건
a와 b가 같은 경우는 둘 중 아무 수나 리턴하세요.
a와 b는 -10,000,000 이상 10,000,000 이하인 정수입니다.
a와 b의 대소관계는 정해져있지 않습니다.
입출력 예
a	b	return
3	5	12
3	3	3
5	3	12
"""


def solution(a, b):
    # 0.006097916599999999
    # return sum(range(a, b+1) if a < b else range(b, a+1))

    # code refactoring
    # 0.003527229800000001
    return (abs(a-b)+1) * (a+b) // 2

print(solution(3, 5))
print(solution(3, 3))
print(solution(5, 3))
print(solution(5, -3))
print(solution(-5, -3))

import timeit
avg_time = 0.
tests = [[3, 5], [3, 3], [5, 3], [5, -3], [-5, -3]]
for t in tests:
    avg_time += timeit.timeit(lambda: solution(*t), number=10000)
print(f'avg_time: {avg_time / len(tests)}')

File: lev1/두정수사이의합.c
#include <stdio.h>
#include <stdbool.h>
#include <stdlib.h>
#include <time.h>

long long solution(int a, int b) {
    // // running time: 0.847277
    // long long ans = 0;
    // if(b < a){
    //     int t = a;
    //     a = b;
    //     b = t;
    // }
    // for(int i=a; i<=b; i++)
    //     ans += i;
    // return ans;

    // code refactoring, sum of arithmetic sequence - 0.000511
    if(a==b) return a;
    
    int n = a < b ? b-a+1 : a-b+1;
    int sum = a+b;
    return (long long)n*sum/2;
}


int main(){
    clock_t tic = clock();
    for (int i=0; i<10000; i++){
        solution(3, 5);
        solution(3, 3);
        solution(5, 3);
        solution(0, 10000); 
        solution(-10000, 0);
        solution(-10000, 10000);
    }
    clock_t toc = clock();
    printf("running time: %f\n", (double)(toc-tic) / CLOCKS_PER_SEC);
    return 0;
}